### PR TITLE
Observe with Terraform Import when Observe Only

### PR DESCRIPTION
**Pull request by turkenh**
### Description of your changes

This PR add support Observe Only management policy to the upjet common controller.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

By following the procedure [here](https://gist.github.com/turkenh/a1641a8e0253bdaa63613aba0b39cee9) and observing some resources, AWS VPC, AWS EKS Cluster ...

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: pkg/pipeline/run.go [hunk 1]
```diff
@@ -78,7 +78,12 @@ func Run(pc *config.Provider, rootDir string) { // nolint:gocyclo
 					Resource:           resources[name],
 					ParametersTypeName: paramTypeName,
 				})
-				ctrlPkgPath, err := ctrlGen.Generate(resources[name], versionGen.Package().Path())
+				// Note(turkenh): With the following, we are assuming that the
+				// generated provider will have a features package under
+				// internal/features. This is a reasonable assumption for now
+				// but we may want to make it configurable in the future.
+				featuresPkgPath := fmt.Sprintf("%s/internal/features", pc.ModulePath)
+				ctrlPkgPath, err := ctrlGen.Generate(resources[name], versionGen.Package().Path(), featuresPkgPath)
```
Comment (turkenh) [inline pkg/pipeline/run.go:86]: Note to myself: We may no longer need feature flag at upjet level.

Comment (turkenh) [reply pkg/pipeline/run.go:86]: Ok, remembered, we need this while generating the controller.go: 

```go
if o.Features.Enabled({{ .FeaturesPackageAlias }}EnableAlphaManagementPolicies) {
		opts = append(opts, managed.WithManagementPolicies())
	}
```

File: pkg/controller/external.go [hunk 2]
```diff
@@ -110,10 +111,67 @@ func (e *external) Observe(ctx context.Context, mg xpresource.Managed) (managed.
...
+			}, nil
+		}
+		// If the resource doesn't exist, we don't need to do anything else.
+		// We report it to the managed reconciler as a non-existent resource and
+		// it will take care of reporting it to the user as an error case for
+		// observe-only policy.
+		if !res.Exists {
+			return managed.ExternalObservation{
+				ResourceExists: false,
+			}, nil
```
Comment (sergenyalcin) [inline pkg/controller/external.go:145]: I am asking to make sure I understand correctly. We do not return an error here if the resource does not exist. As I understand, In this type of case, the responsibility of returning/reporting the error to the user will be on the managed reconciler side, right?

**Review by sergenyalcin — COMMENTED** [review]
Thanks @turkenh I asked a question

File: pkg/controller/external.go (hunk 2 above)
Comment (turkenh) [reply pkg/controller/external.go:145]: Correct: https://github.com/crossplane/crossplane-runtime/blob/4e1673b7141f7449c4ebeea11fa86a184025390b/pkg/reconciler/managed/reconciler.go#L820-L826

**Comment by turkenh** [comment]
Rebased on latest main.

**Review by ulucinar — COMMENTED** [review]
Thank you @turkenh, left some comments for you to consider.

File: pkg/terraform/workspace.go [hunk 3]
```diff
@@ -346,6 +349,63 @@ func (w *Workspace) Plan(ctx context.Context) (PlanResult, error) {
...
+	// provider case, and we simply return does not exist in this case.
+	if len(w.terraformID) == 0 {
+		return ImportResult{
+			Exists: false,
+		}, nil
+	}
+
+	// Note(turkenh): We remove the state file since the import command wouldn't work if tfstate contains
+	// the resource already.
+	if err := w.fs.Remove(filepath.Join(w.dir, "terraform.tfstate")); err != nil {
```
[the remark this answers is not in the document]
Comment (turkenh) [reply pkg/terraform/workspace.go:377]: Good catch! 
Ignored if not exist.

File: pkg/terraform/workspace.go [hunk 4]
```diff
@@ -346,6 +349,63 @@ func (w *Workspace) Plan(ctx context.Context) (PlanResult, error) {
...
+	raw, err := w.fs.ReadFile(filepath.Join(w.dir, "terraform.tfstate"))
+	if err != nil {
+		return ImportResult{}, errors.Wrap(err, "cannot read terraform state file")
+	}
+	s := &json.StateV4{}
+	if err := json.JSParser.Unmarshal(raw, s); err != nil {
+		return ImportResult{}, errors.Wrap(err, "cannot unmarshal tfstate file")
+	}
+	return ImportResult{
+		Exists: s.GetAttributes() != nil,
```
[the remark this answers is not in the document]
Comment (turkenh) [reply pkg/terraform/workspace.go:404]: Correct. I was on the fence but ended up adding this to be defensive as you noticed.

File: pkg/terraform/workspace.go [hunk 5]
```diff
@@ -346,6 +349,63 @@ func (w *Workspace) Plan(ctx context.Context) (PlanResult, error) {
...
+	}
+
+	out, err := w.runTF(ctx, ModeSync, "import", "-input=false", "-lock=false", fmt.Sprintf("%s.%s", tr.GetTerraformResourceType(), tr.GetName()), w.terraformID)
+	w.logger.Debug("import ended", "out", w.filterFn(string(out)))
+	if err != nil {
+		// Note(turkenh): This is not a great way to check if the resource does not exist, but it is the only
+		// way we can do it for now. Terraform import does not return a proper exit code for this case or
+		// does not support -json flag to parse the returning error in a better way.
+		// https://github.com/hashicorp/terraform/blob/93f9cff99ffbb8d536b276a1be40a2c45ca4a67f/internal/terraform/node_resource_import.go#L235
+		if strings.Contains(string(out), "Cannot import non-existent remote object") {
```
[the remark this answers is not in the document]
Comment (turkenh) [reply pkg/terraform/workspace.go:388]: Probably, but I didn't feel good about reading/parsing terraform state after getting an error to my import command.

File: pkg/pipeline/run.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (turkenh) [reply pkg/pipeline/run.go:86]: Just pushed https://github.com/upbound/upjet/pull/168/commits/462dd33a1cb827c8785378201fbe801726291186 for backward compatibility as you suggested.

**Review by ulucinar — APPROVED** [review]
Thanks @turkenh, lgtm.

**Pull request merged**